pkg/cloud/vsphere/services/govmomi/service.go

```diff
@@ -135,19 +135,17 @@ func (vms *VMService) DestroyVM(ctx *context.MachineContext) (infrav1.VirtualMac
 		return vm, err
 	}
 
-	// check if VM actually exists
-	if ctx.VSphereMachine.Spec.MachineRef != "" {
-		moRefID, err := findVMByInstanceUUID(ctx)
-		if err != nil {
-			return vm, err
-		}
-		if moRefID == "" {
-			// No vm exists
-			// remove the MachineRef and set the vm state to notfound
-			ctx.VSphereMachine.Spec.MachineRef = ""
-			vm.State = infrav1.VirtualMachineStateNotFound
-			return vm, nil
-		}
+	// check if the vm existts
+	moRefID, err := findVMByInstanceUUID(ctx)
```

Comment:
Does findVMByInstanceUUID return non-nil err if instance is not found?

Reply:
yes that is the behavior
https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/20c489391b13de471fb5e86e5e9de3621dc77982/pkg/cloud/vsphere/services/govmomi/util.go#L44-L55

Reply:
If so, then the early return in the next line seems like the problem?

Reply:
Right, but `FindByInstanceUUID` can still return non-nil error if not found?

Reply:
Oh, you are saying it does return non-nil error if not found

Reply:
yup

Reply:
> If so, then the early return in the next line seems like the problem?

Which early return?

Reply:
```
moRefID, err := findVMByInstanceUUID(ctx)
if err != nil {
      return vm, err // THIS ONE
}
if moRefID == "" {
    // No vm exists
    // remove the MachineRef and set the vm state to notfound
    ctx.VSphereMachine.Spec.MachineRef = ""    // THIS NEVER HAPPENS
    vm.State = infrav1.VirtualMachineStateNotFound
    return vm, nil
}
```

^ if err is not nil from findVMByInstanceUUID when a VM doesn't exist, then we never end up unset MachineRef

Reply:
The fact that it used to work tells me no error is returned if the item isn't found.

Reply:
No error is returned by `findVMByInstanceUUID` if the VM is simply not found. An error is only returned if there was an error finding the VM. See https://github.com/vmware/govmomi/blob/f01505d273e85c71261f91457191e65fcb971be3/object/search_index.go#L122-L144 for the call that `findVMByInstanceUUID` makes.

This is tested, existing code. We're fine.

Reply:
> The fact that it used to work tells me no error is returned if the item isn't found.

I didn't know this was the case, which version do we know this was working? Btw it's not machine delete that's not working, it's machine delete on an already deleted VM.

Reply:
Hi @yastij,

I think @andrewsykim raises a good point. Let's add a comment summarizing the above behavior to **our** code so it's clear why `findByInstanceUUID` doesn't return an error if the VM isn't found.

Reply:
> The fact that it used to work tells me no error is returned if the item isn't found.

I'm specifically referring to https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/release-0.3/pkg/cloud/vsphere/services/govmomi/util.go#L171-L233.

Reply:
`methods.FindByUuid(ctx, s.c, &req)` calls [RoundTrip](https://github.com/vmware/govmomi/blob/master/vim25/soap/client.go#L535-L606) which has this snippet of code 
```
switch res.StatusCode {
case http.StatusOK:
 // OK
case http.StatusInternalServerError:
// Error, but typically includes a body explaining the error
default:
    return errors.New(res.Status)
}
```
there's no case for a 404 which was why I was asking but I'm not sure if govmomi is expected to return 404 for this request anyways

Reply:
Hi @andrewsykim,

This activity diagram is the best way I know to explain the old model:

![machine-reconcile](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-vsphere/master/docs/design/machine-controller-reconcile.svg?sanitize=true)

Reply:
@dougm  do you know if SearchIndex `FindByUuid` returns a non-nil err if the VM doesn't exist?

Reply:
Hi @andrewsykim,

> there's no case for a 404

Because you're not accessing a VM with the call, you're accessing the search index, which will never 404 unless the SDK server has a massive meltdown. That's the point of the MoRef. It's the basis of a functional API. Most functions are linked to services, ex. SearchIndex, and the first param of most calls is a MoRef used to indicate the object that is the subject of the operation.

Reply:
Thanks I was missing that context

Reply:
FWIW, please see https://pubs.vmware.com/vi3/sdk/ReferenceGuide/vim.SearchIndex.html#findByUuid

> The virtual machine or host managed entity that is found. If no managed entities are found, null is returned. Only a single entity is returned, even if there are multiple matches.

If no object is found it isn't an error -- it just means a nil value is returned by that function. Which is good, because otherwise it would mean we'd need to check for a specific error. Although I suppose that at least is more obvious and in-line with Go. I wonder if that makes more sense? @dougm, what do you think? I suppose a change like that though could have massive consequences for existing code that uses the search index via govmomi.